# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/windows/sdn/sdnTextAccessible.cpp
# Commit: a53fb045a8e3
# Full Hash: a53fb045a8e34c1e1f14ca7b705eb9a5a0e76db2
# Author: James Teh <jteh@mozilla.com>
# Date: 2021-05-02 09:29:52
# Regressor Bug: 1694865
# File Overlap Count: 1
# Description:
#   Bug 1694865 part 19: Make EnumVariant, GeckoCustom and sdn*Accessible classes aggregate MsaaAccessible/MsaaDocAccessible instead of AccessibleWrap/DocAccessibleWrap. r=morgan
#   
#   These objects need to aggregate the primary COM object (MsaaAccessible).
#   Once the IUnknown implementation moves out of AccessibleWrap, it won't be possible to aggregate *AccessibleWrap any more.
#   
# ==============================================================================

diff -r 5bd374b642fe -r a53fb045a8e3 accessible/windows/sdn/sdnTextAccessible.cpp
--- a/accessible/windows/sdn/sdnTextAccessible.cpp	Sat May 01 21:09:04 2021 +0000
+++ b/accessible/windows/sdn/sdnTextAccessible.cpp	Sat May 01 22:29:25 2021 +0000
@@ -28,18 +28,19 @@
 
 IMPL_IUNKNOWN_QUERY_HEAD(sdnTextAccessible)
 IMPL_IUNKNOWN_QUERY_IFACE(ISimpleDOMText)
-IMPL_IUNKNOWN_QUERY_TAIL_AGGREGATED(mAccessible)
+IMPL_IUNKNOWN_QUERY_TAIL_AGGREGATED(mMsaa)
 
 STDMETHODIMP
 sdnTextAccessible::get_domText(BSTR __RPC_FAR* aText) {
   if (!aText) return E_INVALIDARG;
   *aText = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  AccessibleWrap* acc = mMsaa->LocalAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
   nsAutoString nodeValue;
 
-  mAccessible->GetContent()->GetNodeValue(nodeValue);
+  acc->GetContent()->GetNodeValue(nodeValue);
   if (nodeValue.IsEmpty()) return S_FALSE;
 
   *aText = ::SysAllocStringLen(nodeValue.get(), nodeValue.Length());
@@ -55,7 +56,7 @@
                                             &width, &height);
   if (FAILED(rv)) return rv;
 
-  DocAccessible* document = mAccessible->Document();
+  DocAccessible* document = mMsaa->LocalAcc()->Document();
   NS_ASSERTION(
       document,
       "There must always be a doc accessible, but there isn't. Crash!");
@@ -80,9 +81,10 @@
   if (!aX || !aY || !aWidth || !aHeight) return E_INVALIDARG;
   *aX = *aY = *aWidth = *aHeight = 0;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  AccessibleWrap* acc = mMsaa->LocalAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
-  nsIFrame* frame = mAccessible->GetFrame();
+  nsIFrame* frame = acc->GetFrame();
   NS_ENSURE_TRUE(frame, E_FAIL);
 
   nsPoint startPoint, endPoint;
@@ -103,7 +105,7 @@
     sum.UnionRect(sum, rect);
   }
 
-  nsPresContext* presContext = mAccessible->Document()->PresContext();
+  nsPresContext* presContext = acc->Document()->PresContext();
   *aX = presContext->AppUnitsToDevPixels(sum.X());
   *aY = presContext->AppUnitsToDevPixels(sum.Y());
   *aWidth = presContext->AppUnitsToDevPixels(sum.Width());
@@ -115,18 +117,16 @@
 STDMETHODIMP
 sdnTextAccessible::scrollToSubstring(unsigned int aStartIndex,
                                      unsigned int aEndIndex) {
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  AccessibleWrap* acc = mMsaa->LocalAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
-  RefPtr<nsRange> range = nsRange::Create(mAccessible->GetContent());
-  if (NS_FAILED(range->SetStart(mAccessible->GetContent(), aStartIndex)))
-    return E_FAIL;
+  RefPtr<nsRange> range = nsRange::Create(acc->GetContent());
+  if (NS_FAILED(range->SetStart(acc->GetContent(), aStartIndex))) return E_FAIL;
 
-  if (NS_FAILED(range->SetEnd(mAccessible->GetContent(), aEndIndex)))
-    return E_FAIL;
+  if (NS_FAILED(range->SetEnd(acc->GetContent(), aEndIndex))) return E_FAIL;
 
   nsresult rv = nsCoreUtils::ScrollSubstringTo(
-      mAccessible->GetFrame(), range,
-      nsIAccessibleScrollType::SCROLL_TYPE_ANYWHERE);
+      acc->GetFrame(), range, nsIAccessibleScrollType::SCROLL_TYPE_ANYWHERE);
   return GetHRESULT(rv);
 }
 
@@ -135,9 +135,10 @@
   if (!aFontFamily) return E_INVALIDARG;
   *aFontFamily = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  AccessibleWrap* acc = mMsaa->LocalAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
-  nsIFrame* frame = mAccessible->GetFrame();
+  nsIFrame* frame = acc->GetFrame();
   if (!frame) return E_FAIL;
 
   RefPtr<nsFontMetrics> fm = nsLayoutUtils::GetFontMetricsForFrame(frame, 1.0f);